### PR TITLE
SBERDOMA-969 speed up analytics report query

### PR DESCRIPTION
**Pull request by sitozzz**
- [x] Build query with knex (parse default data format)
- [x] Add return type to `GQLCustomSchema`

### CODE REVIEW

File: apps/condo/pages/reports/detail/report_by_tickets/index.tsx [hunk 1]
```diff
@@ -277,6 +286,7 @@ const TicketAnalyticsPage: ITicketAnalyticsPage = () => {
                                 },
                             })
                         })
+                        console.log(series)
```
Comment (dkoviazin) [inline apps/condo/pages/reports/detail/report_by_tickets/index.tsx:289]: console.log

File: apps/condo/domains/ticket/constants/common.js [hunk 2]
```diff
@@ -1,9 +1,11 @@
 const MAX_TICKET_REPORT_COUNT = 1000
 const PDF_REPORT_WIDTH = 1600
 const DATE_DISPLAY_FORMAT = 'DD.MM.YYYY'
+const TICKET_REPORT_SPECIFICATIONS = ['day', 'week']
```
Comment (dkoviazin) [inline apps/condo/domains/ticket/constants/common.js:4]: specifications is not an informative variable name

**Review by dkoviazin — APPROVED** [review]

**Review by AntonAL — COMMENTED** [review]
It's too complex to figure out the implementation.
Please, write a comments about what `AnalyticsQueryBuilder` does in common.

File: apps/condo/domains/common/utils/serverSchema/AnalyticsQueryBuilder.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
+const has = require('lodash/has')
+
+class AnalyticsQueryBuilder {
```
Comment (AntonAL) [inline apps/condo/domains/common/utils/serverSchema/AnalyticsQueryBuilder.js:3]: Please, write comments about what this class does in common and provide usage examples.

File: apps/condo/domains/ticket/schema/TicketAnalyticsReportService.test.js [hunk 4]
```diff
@@ -2,22 +2,19 @@
...
 const { makeClientWithProperty } = require('@condo/domains/property/utils/testSchema')
 const { createTestTicket, makeClientWithTicket } = require('@condo/domains/ticket/utils/testSchema')
-const { TICKET_STATUS_TYPES } = require('@condo/domains/ticket/constants')
 const { makeClient } = require('@core/keystone/test.utils')
-const DATE_FORMAT = 'DD.MM.YYYY'
-const NOW_DATE = moment().format(DATE_FORMAT)
 
 describe('TicketAnalyticsReportService', () => {
     describe('User', () => {
         it('can read TicketAnalyticsReportService grouped counts [day, status]', async () => {
```
Comment (AntonAL) [inline apps/condo/domains/ticket/schema/TicketAnalyticsReportService.test.js:13]: Could you please write a test case with more than one ticket and inspect count of grouped result items.

**Comment by sitozzz** [comment]
Added docstrings & rename service class util

File: apps/condo/domains/ticket/schema/TicketAnalyticsReportService.test.js (hunk 4 above)
Comment (sitozzz) [reply apps/condo/domains/ticket/schema/TicketAnalyticsReportService.test.js:13]: @AntonAL Done

File: apps/condo/domains/common/utils/serverSchema/AnalyticsQueryBuilder.js (hunk 3 above)
Comment (sitozzz) [reply apps/condo/domains/common/utils/serverSchema/AnalyticsQueryBuilder.js]: @AntonAL added some docstrings. Could you check this?

File: apps/condo/domains/common/utils/serverSchema/GqlToKnexBaseAdapter.js [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    where = []
+    dayGroups = ['day', 'month', 'week', 'quarter', 'year']
+    dayGroup = 'day'
+    dateRange = { from: null, to: null }
+    groups = []
+    result = null
+
+    /**
+     * Formatting GQL expressions to Knex data structure
+     * Split up where conditions by structure described at https://knexjs.org/#Builder-where
```
Comment (AntonAL) [inline apps/condo/domains/common/utils/serverSchema/GqlToKnexBaseAdapter.js:14]: I think, we should document it more precisely:

```suggestion
     * Converts Keystone GraphQL `…WhereInput` conditions to shape, that will be used by working with Knex methods, described at https://knexjs.org/#Builder-where
```

File: apps/condo/domains/ticket/schema/TicketAnalyticsReportService.js [hunk 6]
```diff
@@ -21,55 +47,60 @@ const TicketAnalyticsReportService = new GQLCustomSchema('TicketAnalyticsReportS
...
+                            }
+                        }
+                        return {
+                            ...searchResult,
+                            dayGroup: moment(dayGroup).format(DATE_DISPLAY_FORMAT),
+                            count:parseInt(count),
+                        }
+                    }).sort((a, b) =>
+                        moment(a.dayGroup, DATE_DISPLAY_FORMAT).format('X') - moment(b.dayGroup, DATE_DISPLAY_FORMAT).format('X'))
                 return { result }
```
Comment (AntonAL) [inline apps/condo/domains/ticket/schema/TicketAnalyticsReportService.js:104]: You will have `data.result.result` in output, let it be `data.result.groups`, for example.

```suggestion
                return { groups: result }
```

File: apps/condo/domains/ticket/schema/TicketAnalyticsReportService.js [hunk 7]
```diff
@@ -21,55 +47,60 @@ const TicketAnalyticsReportService = new GQLCustomSchema('TicketAnalyticsReportS
         },
         {
             access: true,
-            type: 'type TicketAnalyticsReportOutput { result: JSON! }',
+            type: 'type TicketAnalyticsReportOutput { result: [TicketGroupedCounter!] }',
```
Comment (AntonAL) [inline apps/condo/domains/ticket/schema/TicketAnalyticsReportService.js:50]: You will have `data.result.result` in output, let it be `data.result.groups`, for example.

```suggestion
            type: 'type TicketAnalyticsReportOutput { groups: [TicketGroupedCounter!] }',
```

File: apps/condo/domains/ticket/utils/serverSchema/analytics.helper.js [hunk 8]
```diff
@@ -41,82 +26,59 @@ const sortStatusesByType = (statuses) => {
...
-        translates,
+        if (whereIn[0].length && whereIn[1].length) {
+            this.result = await query.groupBy(this.aggregateBy)
+                .where(where.reduce((acc, current) => ({ ...acc, ...current }), {}))
+                .whereIn(whereIn[0], whereIn[1])
+                .whereBetween('createdAt', [this.dateRange.from, this.dateRange.to])
+        }
+        this.result = await query.groupBy(this.aggregateBy)
+            .where(where.reduce((acc, current) => ({ ...acc, ...current }), {}))
+            .whereBetween('createdAt', [this.dateRange.from, this.dateRange.to])
```
[the remark this answers is not in the document]
Comment (sitozzz) [reply apps/condo/domains/ticket/utils/serverSchema/analytics.helper.js]: Already solved this

File: apps/condo/domains/ticket/schema/TicketAnalyticsReportService.js (hunk 7 above)
Comment (sitozzz) [reply apps/condo/domains/ticket/schema/TicketAnalyticsReportService.js]: Done

File: apps/condo/domains/ticket/schema/TicketAnalyticsReportService.js (hunk 6 above)
Comment (sitozzz) [reply apps/condo/domains/ticket/schema/TicketAnalyticsReportService.js]: Done

File: apps/condo/domains/common/utils/serverSchema/GqlToKnexBaseAdapter.js (hunk 5 above)
Comment (sitozzz) [reply apps/condo/domains/common/utils/serverSchema/GqlToKnexBaseAdapter.js]: updated comment according to your suggestion

**Review by AntonAL — APPROVED** [review]

**Pull request merged**